Sources/SwiftASN1/Basic ASN1 Types/ObjectIdentifier.swift

```diff
@@ -143,25 +144,65 @@ extension ASN1ObjectIdentifier: Hashable {}
 
 extension ASN1ObjectIdentifier: Sendable {}
 
-extension ASN1ObjectIdentifier: ExpressibleByArrayLiteral {
+extension ASN1ObjectIdentifier {
+    /// Initializes ``ASN1ObjectIdentifier`` from its OID components
+    /// - Parameter elements: The OID components
     @inlinable
-    public init(arrayLiteral elements: UInt...) {
+    public init(elements: some Collection<UInt>) throws {
         var bytes = [UInt8]()
-        var components = elements[...]
-        guard let firstComponent = components.popFirst(), let secondComponent = components.popFirst() else {
-            preconditionFailure("Invalid number of OID components: must be at least two!")
+
+        guard elements.count >= 2 else {
+            throw ASN1Error.tooFewOIDComponents(
+                reason: "Invalid number of OID components: must be at least two!"
+            )
         }
 
-        let serializedFirstComponent = (firstComponent * 40) + secondComponent
+        var iterator = elements.makeIterator()
+
+        // Safe to force unwrap first two elements as length >= 2 check has been performed above
+        let serializedFirstComponent = (iterator.next()! * 40) + iterator.next()!
         ASN1ObjectIdentifier._writeOIDSubidentifier(serializedFirstComponent, into: &bytes)
 
-        while let component = components.popFirst() {
+        while let component = iterator.next() {
             ASN1ObjectIdentifier._writeOIDSubidentifier(component, into: &bytes)
         }
         self.bytes = bytes[...]
     }
 }
 
+extension ASN1ObjectIdentifier: ExpressibleByStringLiteral {
+    @inlinable
+    public init(stringLiteral dotRepresentation: String) {
+        // To allow for invalid strings to be tested, parsing is performed in a separate initializer that  `throws`
+        // (this initializer conforms to ExpressibleByStringLiteral so cannot throw)
+        try! self.init(dotRepresentation: dotRepresentation)
+    }
+
+    @inlinable
+    init(dotRepresentation: String) throws {
```

Comment:
We should also add a variant that takes `Substring`, and implement this one in terms of that one.